isce2_topsapp/solid_earth_tides.py

```diff
@@ -6,96 +6,382 @@
+    """
+    state_vectors = []
+    for orbit_xml in orbit_xmls:
+        state_vectors.extend(get_state_vector_arrays(orbit_xml))
+    orb = Orbit()
+    orb.configure()
+    window_min = slc_start_time - datetime.timedelta(seconds=pad_in_seconds)
+    window_max = slc_start_time + datetime.timedelta(seconds=pad_in_seconds)
+    for sv in state_vectors:
+        if (sv.time < window_max) and (sv.time > window_min):
+            orb.addStateVector(sv)
+    return orb
+
+
+def get_start_time_from_slc_id(slc_id: str) -> pd.Timestamp:
+    acq_start_time_token = slc_id.split('_')[5]
+    return pd.to_datetime(acq_start_time_token)
 
-    return solidtide_corr_ds
 
+def compute_solid_earth_tide_from_gunw(*,
+                                       gunw_path: str,
+                                       reference_or_secondary: str,
+                                       orbit_xmls: list[Path]) -> xr.Dataset:
+    """
+    Read GUNW and compute/export differential SET
 
-def get_gunw_attrs(gunw_path: str) -> Tuple[float, np.array, np.array,
-                                            np.array, xr.Dataset, dict]:
-    """ Access necessary GUNW attributes to compute SET """
+    Parameters
+    ----------
+    gunw_path : str
+        Netcdf of the GUNW
+    reference_or_secondary : str
+        Needs to be either "reference" or "secondary" - refers to category of
+        pass for which SET will be calculated
+    orbit_xml : Path
+        The ESA burst xml
 
+    Returns
+    -------
+    xr.Dataset
+        SET data
+
+    Raises
+    ------
+    ValueError
+        When reference_or_secondary is not specified correctly
+    """
+    if reference_or_secondary not in ['reference', 'secondary']:
+        raise ValueError('acq_type must be in "reference" or "secondary"')
+
+    # get GUNW attributes
     group = 'science/radarMetaData'
     with xr.open_dataset(gunw_path, group=group) as ds:
         wavelength = ds['wavelength'].item()
 
+    group = 'science/radarMetaData/inputSLC'
+    with xr.open_dataset(gunw_path, group=f'{group}/{reference_or_secondary}') as ds:
+        slc_ids = ds['L1InputGranules'].data
+        # Ensure non-empty and sorted by acq_time
+        slc_ids = sorted(list(filter(lambda x: x, slc_ids)))
+        slc_start_time = get_start_time_from_slc_id(slc_ids[0])
+
     group = 'science/grids/imagingGeometry'
-    with xr.open_dataset(gunw_path, group=group, engine='rasterio') as ds:
-        z_meta = ds.heightsMeta.data
-        gt = ds.rio.transform()
-        # convert angles to rad
+    with xr.open_dataset(gunw_path, group=group, mode='r') as ds:
+        # latitude resolution will match the rasterio transformed; i.e. it will be negative
+        lon_res, lat_res = ds.rio.resolution()
+        # We make the resolution positive
+        lat_res = -lat_res
+
+        height_coord_arr = ds.heightsMeta.data
+        # half pixel shift to ensure AREA (i.e. upper left corner) convention assumed by pysolid
```

Comment:
it think calculation of SET should be done for the pixel centroid rather than corner, not sure how pysolid calcs this as it accepts AREA style indeed. Is there anywhere in the code where pysolid does half pixel shift in both directions

Reply:
Not totally clear on what you are asking here.  

The python wrappers for the fortran code indeed want the `AREA` formatted metadata (as you mention) which would align with mintpy's attribute `dict`[https://github.com/insarlab/PySolid/blob/main/src/pysolid/grid.py#L30-L33]. I am assuming the georeference/gdal metadata of the outputs would align with those of the inputs (maybe this is where your question is?). This pysolid notebook is pretty clear about the inputs (not about the outputs): https://github.com/insarlab/PySolid/blob/main/docs/plot_grid_SET.ipynb

Also the other GUNW layers specifically the geometry rasters follow the convention:

1. Gdal metadata is `AREA`
2. The coordinates are `PIXEL`
3. The raster is consistent with 1.

This is what I have seen.